27_11_2019/SimpleImageProcessing/ppImageUtils.cpp

```diff
@@ -2,6 +2,91 @@
 using namespace std;
 
 
+
+Matrix EdgeDetection()
+{
+	Matrix ed(3, 3);
+	for (int i = 0; i < 3; i++)
+		for (int j = 0; j < 3; j++)
+			ed(i, j) = -1;
+
+	ed(1, 1) = 8;
+
+	return ed;
+}
+
+Matrix Sharpen()
+{
+	Matrix sh(3, 3, -1);
+	sh(0, 0) = 0;
+	sh(3 - 1, 1 - 1) = 0;
+	sh(1 - 1, 3 - 1) = 0;
+	sh(3 - 1, 3 - 1) = 0;
+	sh(2 - 1, 2 - 1) = 5;
+
+
+
+	return sh;
+}
+
+Matrix Blur()
+{
+	Matrix bl(3, 3, 1.0 / 9.0);
+
+	return bl;
+}
+
+void check(Matrix &input)
+{
+	if (input.numberOfCols() < 3 || input.numberOfRows() < 3)
+		throw std::runtime_error("Minimum rows or columns should be 3, check input matrix");
+}
+
+Matrix masked_image_generator(Matrix & input, Matrix & convolution)
+{
+	Matrix res(input.numberOfRows() - 2, input.numberOfCols() - 2, 0);
+
+
+	for (int i = 1; i < input.numberOfRows() - 1; ++i)
+		for (int j = 1; j < input.numberOfCols() - 1; j++)
+		{
+			res(i - 1, j - 1) = CalculateConvolution(input, convolution, i, j);
+		}
+
+
+
+	return res;
+
+}
+
+double CalculateConvolution(Matrix & input, Matrix & convolution, int i, int j)
```

Comment:
if input and convolution is not changing here, its better to pass them by const ref, it is a good practice.

Reply:
I tried it, but I am messing it up somehow. Maybe, you can explain me in next tutorial. Also, in the github, how to do stuff like review changes, just push the selected files and things like that. I am not a pro at that so I might need your help.